climada/hazard/tc_surge_bathtub.py

```diff
@@ -0,0 +1,192 @@
+PARTICULAR PURPOSE.  See the GNU Lesser General Public License for more details.
+
+You should have received a copy of the GNU Lesser General Public License along
+with CLIMADA. If not, see <https://www.gnu.org/licenses/>.
+
+---
+
+Define TCSurgeBathtub class.
+"""
+
+__all__ = ['TCSurgeBathtub']
+
+import copy
+import logging
+
+import numpy as np
+import rasterio.warp
+import scipy.sparse as sp
+
+from climada.hazard.base import Hazard
+import climada.util.coordinates as coord_util
+
+LOGGER = logging.getLogger(__name__)
+
+HAZ_TYPE = 'TCSurgeBathtub'
+"""Hazard type acronym for this module"""
+
+MAX_DIST_COAST = 50
+"""Maximum inland distance of the centroids in km."""
+
+MAX_ELEVATION = 10
+"""Maximum elevation of the centroids in m."""
+
+
+class TCSurgeBathtub(Hazard):
+    """TC surge heights in m, a bathtub model with wind-surge relationship and inland decay."""
+
+    def __init__(self):
+        Hazard.__init__(self, HAZ_TYPE)
+
+
+    @staticmethod
+    def from_tc_winds(wind_haz, topo_path, inland_decay_rate=0.2, add_sea_level_rise=0.0):
+        """Compute tropical cyclone surge from input winds.
+
+        Parameters
+        ----------
+        wind_haz : TropCyclone
+            Tropical cyclone wind hazard object.
+        topo_path : str
+            Path to a raster file containing gridded elevation data.
+        inland_decay_rate : float, optional
+            Decay rate of surge when moving inland in meters per km. Set to 0 to deactivate
+            this effect. The default value of 0.2 is taken from Section 5.2.1 of the monograph
+            Pielke and Pielke (1997): Hurricanes: their nature and impacts on society.
+            https://rogerpielkejr.com/2016/10/10/hurricanes-their-nature-and-impacts-on-society/
+        add_sea_level_rise : float, optional
+            Sea level rise effect in meters to be added to surge height.
+        """
+        centroids = copy.deepcopy(wind_haz.centroids)
+
+        if not centroids.coord.size:
+            centroids.set_meta_to_lat_lon()
+
+        # Select wind-affected centroids which are inside MAX_DIST_COAST and |lat| < 61
+        if not centroids.dist_coast.size or np.all(centroids.dist_coast >= 0):
+            centroids.set_dist_coast(signed=True, precomputed=True)
+        coastal_msk = (((wind_haz.intensity > 0).sum(axis=0).A1 > 0)
+                       & (centroids.dist_coast < 0)
+                       & (centroids.dist_coast > -MAX_DIST_COAST * 1000)
+                       & (np.abs(centroids.lat) < 61))
+
+        # Load elevation at coastal centroids
+        coastal_centroids_h = coord_util.read_raster_sample(
+            topo_path, centroids.lat[coastal_msk], centroids.lon[coastal_msk])
+
+        # Update selected coastal centroids to exclude high-lying locations
+        elevation_msk = ((coastal_centroids_h >= 0)
+                         & (coastal_centroids_h <= MAX_ELEVATION + add_sea_level_rise))
+        coastal_msk[coastal_msk] = elevation_msk
```

Comment:
The expression `coastal_msk[coastal_msk] = elevation_msk` looks a bit strange, but it's exactly doing what the comment says: it updates the selected coastal centroids to exclude high-lying locations. If you think this still needs clarification, please say how we would clarify this. For me, it's just the usual way how to update the True entries of a numpy mask.

Reply:
I am also confused by this way of writing it. Why is it not coastal_mask = elevation_msk or coastal_mask = (coastal_msk & elevation_mask) ?

Reply:
The `coastal_centroids_h` information is only obtained for the previously selected centroids. So we only update the previously selected centroids and not those that have been discarded in the previous step. Your proposed alternatives simply don't do the same thing.
```suggestion
        # Update selected coastal centroids to exclude high-lying locations
        elevation_msk = ((coastal_centroids_h >= 0)
                         & (coastal_centroids_h <= MAX_ELEVATION + add_sea_level_rise))
        # Only update the previously selected centroids (for which elevation information has been obtained)
        coastal_msk[coastal_msk] = elevation_msk
```